Fixed bug in cleanup of stale connections

kxk302 opened the pull request:
When releasing a connection that was created 'connection_refresh_time' seconds ago, must clean up the connection before creating a new connection and assigning it to the same variable. Previous call to connection_release() with destroy flag set to true only attempted to remove an already popped connection from the idle set (acted as basically a no-op), and the connection destructor had an erroneous if-clause for the call to disconnect() method, which prevented connection from being cleaned up. This fixes both issues.

trel commented:
Excellent - for bookkeeping we'll create issues for these first, and test and then close them with this commit.

@d-w-moore do we need a test to trap this behavior?

kxk302 commented:
> Excellent - for bookkeeping we'll create issues for these first, and test and then close them with this commit.
> 
> @d-w-moore do we need a test to trap this behavior?

Thanks @trel 
I can create an issue and document the problem. I thought about testing this change but was not sure how. Any testing ideas are appreciated!

trel commented:
Yes, please write up the two issues.  Thanks!

1) The no-op is hard to test, agreed.

2) If you can show that a connection didn't clean up in some way... with an assertion... then the test would fail before the fix...

kxk302 commented:
Created issue: https://github.com/irods/python-irodsclient/issues/341

kxk302 commented:
> Yes, please write up the two issues. Thanks!
> 
> 1. The no-op is hard to test, agreed.
> 2. If you can show that a connection didn't clean up in some way... with an assertion... then the test would fail before the fix...

@trel Updated the connection destructor test. I suppose this is ready for review. Could this possibly make it to v1.1.2 of python-irocsclient? Thanks.

d-w-moore commented:
@kxk302 The only idea I have for the no-op:

```
def test_pool_connection_release_is_not_no_op(self):
    pool = self.sess.pool
    c = pool.get_connection( )
    self.sess.pool.release_connection( c, destroy = True )
    self.assertEqual( pool.idle, set( ) )
    self.assertEqual( pool.active, set( ) )
```

kxk302 commented:
> @kxk302 The only idea I have for the no-op:
> 
> ```
> def test_pool_connection_release_is_not_no_op(self):
>     pool = self.sess.pool
>     c = pool.get_connection( )
>     self.sess.pool.release_connection( c, destroy = True )
>     self.assertEqual( pool.idle, set( ) )
>     self.assertEqual( pool.active, set( ) )
> ```

@d-w-moore The no-op was my mistake. I did not check that release_connection() does not release connection resources. Hence, I replaced it with 'del conn', and with the destructor working correctly now, we should be good. Also, I checked and there are existing tests for idle/active set size after a connection is created/released.

d-w-moore commented:
Fair enough. Perhaps this is a better test anyway. It basically tests that old connections are let go when they age out beyond the refresh_time interval.
```
    def test_auto_release_of_old_connection(self):

        do_something = lambda ses: ses.collections.get(helpers.home_collection(ses))

        def get_conn_identity(pool,unittest_obj=None):
            s = (pool.active | pool.idle)
            if unittest_obj:
                unittest_obj.assertEqual (len(s), 1)
            return [conn for conn in s][0]

        with helpers.make_session(refresh_time = 5) as ses:
            do_something(ses)
            c1 = get_conn_identity (ses.pool,self)
            time.sleep(10)
            do_something(ses)
            c2 = get_conn_identity (ses.pool,self)
            self.assertIsNot(c1, c2)
```

kxk302 commented:
> Fair enough. Perhaps this is a better test anyway. It basically tests that old connections are let go when they age out beyond the refresh_time interval.
> 
> ```
>     def test_auto_release_of_old_connection(self):
> 
>         do_something = lambda ses: ses.collections.get(helpers.home_collection(ses))
> 
>         def get_conn_identity(pool,unittest_obj=None):
>             s = (pool.active | pool.idle)
>             if unittest_obj:
>                 unittest_obj.assertEqual (len(s), 1)
>             return [conn for conn in s][0]
> 
>         with helpers.make_session(refresh_time = 5) as ses:
>             do_something(ses)
>             c1 = get_conn_identity (ses.pool,self)
>             time.sleep(10)
>             do_something(ses)
>             c2 = get_conn_identity (ses.pool,self)
>             self.assertIsNot(c1, c2)
> ```

Thanks  @d-w-moore. This is a great test. I will add it to my branch and update the PR.

d-w-moore commented:
Well, I'm not sure now - I tried this test with a revert of the fix and it still passed.  So, maybe more investigation is needed....

kxk302 commented:
> Well, I'm not sure now - I tried this test with a revert of the fix and it still passed. So, maybe more investigation is needed....

I'll look into it as well.

d-w-moore commented:
@kxk302  Is there a way to reproduce the original issue #341 ? As far as I can tell, the irods.pool.Pool's `release_connection` has sufficient logic to make sure the Connection object being released cleans up and "goes away" (in the sense that Python's garbage collector takes care of calling  Connection.__del__  -- as long as no extra references to the Connection have been squirreled away). 
I've made this development [branch](https://github.com/d-w-moore/python-irodsclient/tree/kxk302_issue_341) for testing. If we can figure out how to reproduce the issue reliably, I could add a test around that and run it (and the whole suite) against this branch

kxk302 commented:
> @kxk302 Is there a way to reproduce the original issue #341 ? As far as I can tell, the irods.pool.Pool's `release_connection` has sufficient logic to make sure the Connection object being released cleans up and goes away (in the sense that Python's garbage collector takes care of calling Connection.__del__ -- as long as no extra references to the Connection have been squirreled away). I've made this development [branch](https://github.com/d-w-moore/python-irodsclient/tree/kxk302_issue_341) for testing. If we can figure out how to reproduce the issue reliably, I could add a test around that and run it (and the whole suite) against this branch

In the original code, when a connection was stale, I would call connection_release() on a connection that was already popped from the idle set, hence, connection_release() would do nothing (a 'no-op'). The problem was I would then immediately create a new connection and assign it to the 'conn' variable. We would lose the reference on the old connection, and connection destructor had a bug, hence, a leak would ensue. I noticed this by eye balling the code (there was no errors, etc that alerted me to this issue). Fixing the destructor bug, and calling the destructor explicitly, will fix the problem for sure.  However, writing a test is tricky and I'm not sure how to go about it. If you like to discuss further, let me know and we can Zoom and discuss.

d-w-moore commented:
@kxk302  I'm fairly open this week for doing a Zoom session about this.  If we can demonstrate the leak manually, I may then be able to write a test.

d-w-moore commented:
FWIW, though, it appears to me, on a superficial read at least, that the logic is still fine.  A `conn` that gets popped does indeed cause the `release_connection` call to be a no-op, but this should be ok in the end because the desired condition for `destroy = True ` in that call is still satisfied.  (Ie. the conn does not get re-added to the active or idle pool, so it will "go away" and __del__ should be called in a timely fashion.)

kxk302 commented:
> FWIW, though, it appears to me, on a superficial read at least, that the logic is still fine. A `conn` that gets popped does indeed cause the `release_connection` call to be a no-op, but this should be ok in the end because the desired condition for `destroy = True ` in that call is still satisfied. (Ie. the conn does not get re-added to the active or idle pool, so it will "go away" because of garbage collection, and **del** should be called in a timely fashion.)

Correct. The release_connection() would do nothing. The fix to destructor means that the unreferenced object would be cleanup now. I just replaced the no-op release_connection() with an explicit call to the destructor, so it's clear we are cleaning up the resources.

d-w-moore commented:
@kxk302 - ok, let me know if you would like to meet this week to determine if the issue is still happening.

d-w-moore commented:
Ah, wait!  I see the confusion.... Reference these two lines of code in the updated `irods/pool.py` :
```
69 del conn
70 conn = Connection(self, self.account)
```
It does what you want, but in truth line 69 is actually redundant in Python's reference-counted scheme, since line 70 will clear the existing `conn ` reference (and call its destructor implicitly) before assigning conn the new value.

So... we actually don't need the explicit `del` call, only the changes within the \_\_del\_\_ method.

d-w-moore commented:
another example of the above
```
>>> class X(object):
...  def __del__(self):
...   print('goodbye')
... 
>>> x=X( )
>>> y = x   # with 2 refs to an object,
>>> del x   # the destructor fires not when del clears x from the namespace, but rather...
>>> y = 1   # ... when the last ref (in y) goes away
goodbye
>>> 

```

kxk302 commented:
@d-w-moore I added a log message to connection destructor, added a test to pool_test.py that ages a connection (so we create a new connection, throw away the old , so that the connection destructor is called). the test re-routes the irods.connection logger to file, and asserts that the connection destructor log message is in the log file.

kxk302 commented:
@d-w-moore  
The new test passes and I've verified that log file has the connection destructor log message. However, when checking the pool_test.py's XML output, I see the below message. Any idea whats causing this?

'''
    <system-err><![CDATA[Exception ignored in: <function Connection.__del__ at 0x10fd76b80>
Traceback (most recent call last):
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 84, in __del__
    self.disconnect()
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 278, in disconnect
    self.send(disconnect_msg)
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 100, in send
    raise NetworkException("Unable to send message")
irods.exception.NetworkException: Unable to send message
]]></system-err>
'''

kxk302 commented:
@d-w-moore 
I updated irods/test/runner.py so that only pools_test.py is run (set 'pattern' to 'pool_test.py'), then ran './run_python_tests.sh'. I then looked up the XMLTestRunner's output file (set it to '/tmp/python-irodsclient/test-reports' in runners.py)

kxk302 commented:
@d-w-moore 
I ran 'python runner.py' and I still see the error. FYI, I have a (dockerized) irods server running locally, that the tests connect to.

d-w-moore commented:
@kxk302 - Is only one test causing the errors? Also, are there any customizations to the iRODS server - rule bases present, plugins installed, etc, that would distinguish it from a vanilla install ?

kxk302 commented:
@d-w-moore All the tests pass. See below. No failures or errors. In System Error section of the file, I see the error message that I posted. The irods server has not rule bases or special plugins, so I think its 'vanilla'. Just curious do you see this error when you run it t your end? Also, what irods server do you point to?
```
<?xml version="1.0" ?>
<testsuite name="pool_test.TestPool-20220214144012" tests="13" time="11.546" failures="0" errors="0">
        <testcase classname="pool_test.TestPool" name="test_connection_create_time" time="0.052"/>
        <testcase classname="pool_test.TestPool" name="test_connection_destructor_called" time="5.121"/>
        <testcase classname="pool_test.TestPool" name="test_destroy_active" time="0.052"/>
        <testcase classname="pool_test.TestPool" name="test_destroy_idle" time="0.048"/>
        <testcase classname="pool_test.TestPool" name="test_get_connection_refresh_time" time="0.001"/>
        <testcase classname="pool_test.TestPool" name="test_get_connection_refresh_time_negative_connection_refresh_field" time="0.000"/>
        <testcase classname="pool_test.TestPool" name="test_get_connection_refresh_time_no_connection_refresh_field" time="0.000"/>
        <testcase classname="pool_test.TestPool" name="test_get_connection_refresh_time_no_env_file_input_param" time="0.000"/>
        <testcase classname="pool_test.TestPool" name="test_get_connection_refresh_time_none_existant_env_file" time="0.000"/>
        <testcase classname="pool_test.TestPool" name="test_no_refresh_connection" time="1.052"/>
        <testcase classname="pool_test.TestPool" name="test_refresh_connection" time="5.117"/>
        <testcase classname="pool_test.TestPool" name="test_release_connection" time="0.052"/>
        <testcase classname="pool_test.TestPool" name="test_release_disconnected" time="0.050"/>
        <system-out><![CDATA[]]></system-out>
        <system-err><![CDATA[Exception ignored in: <function Connection.__del__ at 0x1017d8b80>
Traceback (most recent call last):
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 84, in __del__
    self.disconnect()
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 278, in disconnect
    self.send(disconnect_msg)
  File "/Users/kxk302/workspace/python-irodsclient/irods/connection.py", line 100, in send
    raise NetworkException("Unable to send message")
irods.exception.NetworkException: Unable to send message
]]></system-err>
</testsuite>

```

d-w-moore commented:
@kxk302 I don't see the error on my end; did see system-out and system-err sections but no errors printed out between the XML tags.
I've got a fresh new iRODS install I tested with once, local to the host node.  I tested again going across a short internet distance to an iRODS server on my desktop machine at work.  But, I've seen nothing yet of the type of error you are seeing.
My only thought is that the disconnect call in the Connection destructor is getting an exception thrown, but it's a shame the xmlrunner doesn't let us know what that is ....

kxk302 commented:
I commented out the NetworkException in connection.py's send method (replaced it with raise) and it seems the error is coming out of test_release_disconnected test, the line that calls 'conn.disconnect' in the first with clause.

d-w-moore commented:
@kxk302 Just wondering, but are you testing with SSL enabled?

mvdbeek commented:
> since line 70 will clear the existing `conn` object away (and call its destructor implicitly) before assigning conn the new value.

Sorry to swoop in here, butI wanted to clarify that Python does not guarantee at which point in time an object will be garbage collected, and so at which moment the `__del__` finalizer is called. Even explicitly calling `del` does not guarantee this, though you could call `__del__()` if you know it can be called more than once (if you call it explicitly and during garbage collection)
The normal pattern for cleaning up resources is to do this in a `finally` block or within a context manager's `__exit__` method.

 I don't know if this matters in this case, but I think adhering to the usual pattern is not a bad idea.

d-w-moore commented:
I'm all for making the tests (and, in general, connection behavior) deterministic. If we can't depend on __del__  to be called immediately, that's fine ; it will be called eventually and we'll avoid the undesirable effect of too many connection objects building up.

If a disconnect( ) call throws in a __del__ invocation that occurs "too late", then we'll just discard the NetworkException and assume the connection was dead anyway.

While I do love me a good context manager idiom a la try / yield / finally , I don't think it's worth messing with here.  There are applications (Galaxy's perhaps included) that maintain a session or connection object for minutes or more, and we probably won't find ourselves within the same context when the session (Connection) gets deleted as when it was created.

kxk302 commented:
I think I've found the issue. In pool_test.py's test_release_disconnected (pasted below), conn.disconnect() is called in the with clause, and then in sess.cleanup(), conn.disconnect() is called again, causing the exception. Looking at the code to see whats the best fix.

```
def test_release_disconnected(self):
        with self.sess.pool.get_connection() as conn:
            self.assertEqual(1, len(self.sess.pool.active))
            self.assertEqual(0, len(self.sess.pool.idle))
            conn.disconnect()

        # even though disconnected, gets put into idle
        self.assertEqual(0, len(self.sess.pool.active))
        self.assertEqual(1, len(self.sess.pool.idle))

        # should remove all connections
        self.sess.cleanup()
        self.assertEqual(0, len(self.sess.pool.active))
        self.assertEqual(0, len(self.sess.pool.idle))
```

d-w-moore commented:
Perhaps in this case, we do without the with clause? Just assign conn = self.sess.pool.get_connection( ) since we don't necessarily want the "weak connection release" in the context manager close - ?  We could get rid of the middle two asserts that test that a disconnected socket has landed among the idle set in the pool. Or is that an important aspect of the connection behavior we are properly motivated to test ?

kxk302 commented:
> Perhaps in this case, we do without the `with` clause? Just assign `conn = self.sess.pool.get_connection( ) `since we don't necessarily want both the disconnect ( ) and the "weak connection release" in the context manager close ? We could get rid of the middle two asserts that test that a disconnected socket has landed among the idle set in the pool. Or is that an important aspect of the connection behavior we are properly motivated to test ?

@d-w-moore with clause's exit() only calls the connection release(), which moves the connection from active to idle set. So, I don't think thats the problem. The problem is 2 subsequent calls to conn.disconnect (the second call coming via sess.cleanup()). It seems we have to make conn.disconnect() safely handle disconnected connection.

d-w-moore commented:
Agreed, we could probably just move the logic
```
if self.socket and getattr(self, "_disconnected", False) == False:
```
into disconnect( ) itself.

kxk302 commented:
@d-w-moore I moved the logic to call disconnect() from __del__() to inside disconnect(). I ran the tests and was still seeing an exception in system error. Debugged it and disconnect() calls send(), which in turn calls sendall(). The sendall would throw an exception as the socket was closed (socket.fileno() would return -1). Extended the logic to call disconnect() so any closed socket is excluded. Pool test pass and I do not see anything in the system error. Could you please review this when you get a chance?

d-w-moore commented:
@kxk302 It looks fine. We may be able to tighten up the logic somewhat in disconnect( ), but that may not be so critical.  The change did cause one test to fail in connection_test.py (test_send_failure) but I don't think it's an important one. (It deliberatly runs disconnect twice and expects a NetworkException.  I don't think that is an important test if we are trying to make the 2nd+ disconnect call a no-op. Will probably delete that test :) )  In any case,  I'll run the test suite on it....

kxk302 commented:
> @kxk302 It looks fine. We may be able to tighten up the logic somewhat in disconnect( ), but that may not be so critical. The change did cause one test to fail in connection_test.py (test_send_failure) but I don't think it's an important one. (It deliberatly runs disconnect twice and expects a NetworkException. I don't think that is an important test if we are trying to make the 2nd+ disconnect call a no-op. Will probably delete that test :) ) In any case, I'll run the test suite on it....

Sounds good @d-w-moore!

kxk302 commented:
Created an issue for this PR: https://github.com/irods/python-irodsclient/issues/345

kxk302 commented:
Created a new PR with squashed commits here: https://github.com/irods/python-irodsclient/pull/346
